enclave_build/src/cache.rs

```diff
@@ -0,0 +1,219 @@
+            schema_version: 2,
+            manifests: Vec::new(),
+        }
+    }
+}
+
+fn schema_default() -> u32 {
+    2
+}
+
+/// Struct which provides operations with the local cache.
+///
+/// The index file is located in the cache root folder and keeps track of images stored in cache.
+///
+/// The cache structure is:
+///
+/// {CACHE_ROOT_PATH}/index.json\
+/// {CACHE_ROOT_PATH}/blobs/sha256\
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_1\
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_2\
+/// etc.
+///
+/// A blob can be:
+///
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/manifest_hash - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/config_hash - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/layer_hash - one of the image layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
+pub struct CacheManager {
+    /// The root folder of the cache
+    root_path: PathBuf,
+
+    /// A map storing the cached images, with the map entry format being (image_reference, image_hash)
+    cached_images: HashMap<String, String>,
+}
+
+impl CacheManager {
+    /// Creates a new CacheManager instance and returns it. As argument, a path to the root folder
+    /// of the cache should be provided.
+    ///
+    /// Apart from that, the function also creates (if not already created) all folders from the path
+    /// specified as argument.
+    ///
+    /// If an index file exists at the path, it loads the file's contents into the 'cached_images'
+    /// field. If not, a new empty index file is created at that path.
+    pub fn new(root_path: &Path) -> Result<Self> {
+        // Create all missing folders, if not already created
+        fs::create_dir_all(&root_path).map_err(EnclaveBuildError::CacheInitError)?;
+
+        // Try to open the index file and read the contents.
+        // If the file is missing, create it.
+        OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(root_path.to_path_buf().join(CACHE_INDEX_FILE_NAME))
+            .map_err(EnclaveBuildError::CacheInitError)?;
+
+        // Load the index content and match image names to config hashes.
+        // If the index file is empty, return an empty hashmap.
+        let cached_images = match Self::fetch_index(root_path) {
+            Ok(index) => index
+                .manifests
+                .iter()
+                // Extract <image_name, manifest_digest> from index
+                .filter_map(|entry| {
+                    entry
+                        .annotations
+                        .get(REF_ANNOTATION)
+                        .map(|img_ref| (img_ref.to_string(), entry.digest.to_string()))
+                })
+                // Read manifest files and map digest to JSON content
+                .filter_map(|(img_ref, manifest_hash)| {
+                    Self::fetch_manifest(root_path, &manifest_hash)
+                        .ok()
```

Comment:
So if the `fetch_*` functions return an error, we discard it or what is happening here?